### PR TITLE
[Bug] Fix NN modules crashing with non-FP32 inputs (reopen #4262)

### PR DESCRIPTION
**Pull request by yaox12**
## Description

Reopen https://github.com/dmlc/dgl/pull/4262.

With this PR, pure-BF16 training converges well. FP16 still requires AMP to get convergence.

Note: the example is attached for awareness and won't be merged.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/graphconv.py [hunk 1]
```diff
@@ -107,8 +107,10 @@ def forward(self, graph, edge_weight):
                                'This leads to square root of zero or negative values.')
 
             dev = graph.device
-            graph.srcdata['_src_out_w'] = th.ones((graph.number_of_src_nodes())).float().to(dev)
-            graph.dstdata['_dst_in_w'] = th.ones((graph.number_of_dst_nodes())).float().to(dev)
+            graph.srcdata['_src_out_w'] = th.ones(
+                    graph.number_of_src_nodes(), dtype=edge_weight.dtype, device=dev)
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/conv/graphconv.py:111]: over 80 char.

File: examples/pytorch/gcn/train_bf16.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--dataset", type=str, default="cora",
+                        help="Dataset name ('cora', 'citeseer', 'pubmed').")
+    args = parser.parse_args()
+    print(f'Training with DGL built-in GraphConv module.')
+ 
+    # load and preprocess dataset
+    transform = AddSelfLoop()  # by default, it will first remove self-loops to prevent duplication
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:65]: over 80 char.

File: examples/pytorch/gcn/train_bf16.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+              . format(epoch, loss.item(), acc))
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--dataset", type=str, default="cora",
+                        help="Dataset name ('cora', 'citeseer', 'pubmed').")
+    args = parser.parse_args()
+    print(f'Training with DGL built-in GraphConv module.')
+ 
+    # load and preprocess dataset
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:64]: For all comments, start with Capital and end with dot.

File: examples/pytorch/gcn/train_bf16.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    # training loop
+    for epoch in range(200):
+        model.train()
+        logits = model(g, features)
+        loss = loss_fcn(logits[train_mask], labels[train_mask])
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        acc = evaluate(g, features, labels, val_mask, model)
+        print("Epoch {:05d} | Loss {:.4f} | Accuracy {:.4f} "
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:54]: it's highly recommended to use f-string in dgl python.

File: examples/pytorch/gcn/train_bf16.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+              . format(epoch, loss.item(), acc))
+
+if __name__ == '__main__':
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:57]: we use double quote in dgl, black should help you fix this in one go.

File: examples/pytorch/gcn/train_bf16.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    # load and preprocess dataset
+    transform = AddSelfLoop()  # by default, it will first remove self-loops to prevent duplication
+    if args.dataset == 'cora':
+        data = CoraGraphDataset(transform=transform)
+    elif args.dataset == 'citeseer':
+        data = CiteseerGraphDataset(transform=transform)
+    elif args.dataset == 'pubmed':
+        data = PubmedGraphDataset(transform=transform)
+    else:
+        raise ValueError('Unknown dataset: {}'.format(args.dataset))
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:73]: ditto.

File: examples/pytorch/gcn/train_bf16.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    model = GCN(in_size, 16, out_size).to(device, torch.bfloat16)
+
+    # model training
+    print('Training...')
+    train(g, features, labels, masks, model)
+    
+    # test the model
+    print('Testing...')
+    acc = evaluate(g, features, labels, masks[2], model)
+    print("Test accuracy {:.4f}".format(acc))
```
Comment (frozenbugs) [inline examples/pytorch/gcn/train_bf16.py:99]: ditto.

File: examples/pytorch/gcn/train_bf16.py (hunk 2 above)
Comment (yaox12) [reply examples/pytorch/gcn/train_bf16.py:65]: This example is for review only and will be deleted before merging.

File: python/dgl/nn/pytorch/conv/graphconv.py (hunk 1 above)
Comment (yaox12) [reply python/dgl/nn/pytorch/conv/graphconv.py]: Updated accordingly.

**Review by BarclayII — APPROVED** [review]

**Pull request merged**